docker/Makefile

```diff
@@ -0,0 +1,119 @@
+# setup (use XXX=<value> make <target> to override)
+include Makefile.image_names
+
+DOCKER_BUILDKIT=1
+
+BUILD_ARGS += \
+	--build-arg FMS_IMAGE=$(FMS_IMAGE) \
+	--build-arg ESMF_IMAGE=$(ESMF_IMAGE) \
+	--build-arg MPI_IMAGE=$(MPI_IMAGE) \
+	--build-arg GT4PY_IMAGE=$(GT4PY_IMAGE) \
+	--build-arg SERIALBOX_IMAGE=$(SERIALBOX_IMAGE) \
+	--build-arg ENVIRONMENT_IMAGE=$(ENVIRONMENT_IMAGE) \
+	--build-arg FORTRAN_IMAGE=$(FORTRAN_IMAGE) \
+	--build-arg WRAPPER_IMAGE=$(WRAPPER_IMAGE) \
+	--build-arg GT4PY_OPTIONALS=$(GT4PY_OPTIONALS) \
+	--build-arg ENV_CUDA_PATH=$(ENV_CUDA_PATH)
+
+# sarus paths
+CORE_TAR=$(TAG_NAME).tar
+CORE_BUCKET_LOC=gs://vcm-jenkins/fv3core_$(CORE_TAR)
+
+CORE_DEPS_IMAGES = $(MPI_IMAGE) $(SERIALBOX_IMAGE) $(GT4PY_IMAGE) $(ENVIRONMENT_IMAGE)
+DEPS_IMAGES ?= $(CORE_DEPS_IMAGES) $(FMS_IMAGE) $(ESMF_IMAGE)
+
+build: fv3core_image
+
+dev_wrapper:
+	if [ -z $(shell docker images -q $(FV3CORE_WRAPPER_IMAGE)) ]; then \
+		$(MAKE) fv3core_wrapper_image; \
+	fi
+	FV3CORE_WRAPPER_IMAGE=$(FV3CORE_WRAPPER_IMAGE) docker-compose run fv3core_wrapper bash
+
+.PHONY: dev fortran_image wrapper_image fv3core_image build build_deps push_deps pull_deps
+
+fortran_image:
+	docker build \
+		$(BUILD_ARGS) \
+		-f $(DEPENDENCIES_DOCKERFILE) \
+		-t $(FORTRAN_IMAGE) \
+		../external/fv3gfs-fortran
+
+# builds only the fortran stages with no external images
+# useful for debugging dockerfile targets in fortran model compilation
+full_fortran_image:
+	FMS_IMAGE=fv3gfs-fms ESMF_IMAGE=fv3gfs-esmf MPI_IMAGE=fv3gfs-mpi $(MAKE) fortran_image
+
+fv3core_image: ## build production container image
+	docker build \
+		$(BUILD_ARGS) \
+		-f $(FV3CORE_DOCKERFILE) \
+		-t $(FV3CORE_IMAGE) \
+		--target=$(FV3CORE_TARGET) \
+		..
+
+fv3core_wrapper_image: fortran_image
+	FV3CORE_TARGET=fv3core_wrapper FV3CORE_IMAGE=$(FV3CORE_WRAPPER_IMAGE) $(MAKE) fv3core_image
+
+build_deps: build_core_deps ## build container images of all dependencies
+	docker build -f $(DEPENDENCIES_DOCKERFILE) -t $(FMS_IMAGE) $(BUILD_ARGS) --target fv3gfs-fms ../external/fv3gfs-fortran
+	docker build -f $(DEPENDENCIES_DOCKERFILE) -t $(ESMF_IMAGE) $(BUILD_ARGS) --target fv3gfs-esmf ../external/fv3gfs-fortran
+
+build_core_deps: ## build container images of dependencies
+	docker build -f $(DEPENDENCIES_DOCKERFILE) -t $(ENVIRONMENT_IMAGE) $(BUILD_ARGS) --target fv3core-environment .
```

Comment:
I appreciate this being separated out! Although in theory docker caching should never rerun these when it has already been done, I keep running into cases where that happens, so appreciate this for now. We may decide later that pulling the image takes just as long as building and reverse our decision, but I like this here. Though, building fv3core still took 3 minutes on my machine :P. So perhaps it was more due to the pip installs than the apt-gets. I am ok with moving on from this for now, as it all seems quite reasonable.

Reply:
I was glad to figure out a way that enabled doing it with a lot of make target re-use! Without it my main concern was too much copy-paste.

Yeah the pip installs do take a while. 3 minutes is similar to the fv3net image. I can put them in the environment image, but it means you have to re-build the environment image if you update the pip installs (including what version of fv3config and mpi4py are being used). I think the current approach will save time compared to that when local image caches are considered, but let me know if we should change it.